Documentation/design-docs/host-probing.md

```diff
@@ -0,0 +1,61 @@
+# .NET Core host probing
+
+The dotnet host uses probing when it searches for actual file on disk for a given asset. This happens for example when it tries to resolve assembly specified in `.deps.json` to a file on disk. The `.deps.json` specifies relative path of the asset in its library and also the relative path to the library (typically package name and version). Consider this small portion of `.deps.json` for example (only relevant parts, everything else omitted):
+```json
+{
+  "targets": {
+    ".NETCoreApp,Version=v2.1": {
+      "Newtonsoft.Json/11.0.2": {
+        "runtime": {
+          "lib/netstandard2.0/Newtonsoft.Json.dll": {}
+        }
+      }
+    }
+  },
+  "libraries": {
+    "Newtonsoft.Json/11.0.2": {
+      "type": "package",
+      "serviceable": true,
+      "path": "newtonsoft.json/11.0.2",
+    }
+  }
+}
+```
+
+The library relative path in this case is `newtonsoft.json/11.0.2` and the asset relative path is `lib/netstandard2.0/Newtonsoft.Json.dll`. So the goal of the probing logic is to find the `Newtonsoft.Json.dll` file using the above relative paths.
+
+## Probing
+The probing itself is done by going over a list of probing paths, which are ordered according to their priority. For each path, the host will append the relative parts of the path as per above and see if the file actually exists on the disk.  
+If the file is found, the probing is done, and the full path just resolved is stored.  
+If the file is not found, the probing continues with the next path on the list.
+If all paths are tried and the asset is still not found this is reported as an error (with the exception of app's `.deps.json` asset, in which case it's ignored).
+
+## Probing paths
```

Comment:
Clarify framework-dependent vs. self-contained?

Reply:
Not exactly sure what you're asking for. I added a paragraph about this at the very end... but it's also already mentioned in the part about Framework directories.